content/blog/2018-06-07-when-to-use-derived-state.md

```diff
@@ -0,0 +1,388 @@
+
+Hopefully it's clear by now why unconditionally overriding state with props is a bad idea. But what if we were to only update the state when the email prop changes? We'll take a look at that pattern next.
+
+### Anti-pattern: Props override state whenever they change
+
+Building on our example above, we could avoid accidentally erasing state by only updating it when `props.email` changes:
+
+```js
+class EmailInput extends Component {
+  state = {
+    prevPropsEmail: this.props.email,
+    email: this.props.email
+  };
+
+  static getDerivedStateFromProps(props, state) {
+    // Any time props.email changes, update state.
+    if (props.email !== state.prevPropsEmail) {
+      return {
+        prevPropsEmail: props.email,
+        email: props.email
+      };
+    }
+    return null;
+  }
+  
+  // ...
+}
+```
+
+We've just made a big improvement. Now our component will no longer erase what we've typed every time it renders. At this point, what we have is fairly similar to the [scroll direction example](blog/2018/03/27/update-on-async-rendering.html#updating-state-based-on-props) we mentioned before. (It will also more closely mirror how `componentWillReceiveProps` would have been used in the past.)
+
+There is still one subtle problem though, and it's probably easiest to illustrate with an example. Let's say our `EmailInput` component is used inside of an "edit profile" form. The first time the form is rendered, the component will display our email address. Let's say we edited the form (including our email address) but then changed our mind and clicked a "reset" button. At this point, we would expect all form fields to return to their initial values– but the `EmailInput` component **will not be reset**. Do you know why?
+
+The reason for this is that `props.email` never actually changed in the above scenario. Both times the "edit profile" form rendered, it would pass our saved email address via props.
+
+This problem could manifest itself even without a "reset" button. For example, imagine a password manager app using the above input component. When navigating between details for two accounts with the same initial email, the input would fail to reset because the prop value passed to the component would be the same! This would be a surprise to the user, as a draft change to one account would appear to affect other accounts that happened to share the same email.
+
+This design is fundamentally flawed, but it's also an easy mistake to make. [I've made it myself.](https://twitter.com/brian_d_vaughn/status/959600888242307072) Fortunately there are two alternatives that work better. The key to both is that **for any piece of state, you need to pick a single component that owns it as the source of truth, and avoid duplicating it in other components.** Let's take a look at each.
```

Comment:
"Let's take a look at each of the alternatives"? After the expansion of the bold text, I'm losing what "each" referred to.